nixos/modules/services/networking/envoy.nix

```diff
@@ -0,0 +1,76 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.envoy;
+
+in
+
+{
+  options.services.envoy = {
+    enable = mkEnableOption "Envoy reverse proxy";
+
+    settings = mkOption {
+      type = (pkgs.formats.json { }).type;
+      default = { };
+      example = literalExpression ''
+        {
+          admin = {
+            access_log_path = "/dev/null";
+            address = {
+              socket_address = {
+                protocol = "TCP";
+                address = "127.0.0.1";
+                port_value = 9901;
+              };
+            };
+          };
+          static_resources = {
+            listeners = [];
+            clusters = [];
+          };
+        }
+      '';
+      description = ''
+        Specify the configuration for Envoy in Nix.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.envoy ];
+    users.groups._envoy = { };
+    users.users._envoy = {
+      description = "Envoy service user";
+      group = "_envoy";
+      isSystemUser = true;
+    };
+    systemd.services.envoy = {
+      description = "Envoy reverse proxy";
+      after = [ "network-online.target" ];
+      requires = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig =
+        let
+          format = pkgs.formats.json { };
+          conf = format.generate "envoy.json" cfg.settings;
+          validateConfig = file:
+            pkgs.runCommand "validate-envoy-conf" { } ''
+              ${pkgs.envoy}/bin/envoy --log-level error --mode validate -c "${file}"
+              ln -s "${file}" "$out"
+            '';
+        in
+        {
+          ExecStart = "${pkgs.envoy}/bin/envoy -c ${validateConfig conf}";
+          User = "_envoy";
+          Group = "_envoy";
+          Restart = "no";
+          CacheDirectory = "envoy";
+          ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
```

Comment:
reloading won't do you much good because the command line of the HUP'd process doesn't change, so it can't pick up the new config file. nixos services are restarted when their unit changes unless it's specifically requested to reload instead, which usually requires config file to be globally accessible.

Reply:
alright removed that